molgenis-js/src/main/java/org/molgenis/js/magma/JsMagmaScriptEvaluator.java

```diff
@@ -79,6 +124,38 @@ public Object eval(String expression, Entity entity, int depth)
 		return value;
 	}
 
+	/**
+	 * Creates magmascript bindings for a given Entity with default depth.
+	 *
+	 * @param entity the entity to bind to the magmascript $ function
+	 * @return Bindings with $ function bound to the entity
+	 */
+	public Bindings createBindings(Entity entity)
+	{
+		return createBindings(entity, ENTITY_REFERENCE_DEFAULT_FETCHING_DEPTH);
+	}
+
+	/**
+	 * Creates magmascript bindings for a given Entity.
+	 *
+	 * @param entity the entity to bind to the magmascript $ function
+	 * @param depth  maximum depth to follow references when creating the entity value map
+	 * @return Bindings with $ function bound to the entity
+	 */
+	public Bindings createBindings(Entity entity, int depth)
```

Comment:
According to IntelliJ this method can be private. I expected this to be called by ```AlgorithmServiceImpl```. Is the algorithm service benefiting from the performance improvements in this PR?

Reply:
There are two methods in the mapping service.

### algorithm editing
While the user is typing on a single algorithm it regularly applies that single algorithm to a couple of entity rows from the appropriate source repository to get a column of target attribute values.

So it runs a single loop over all source repository entities. Reusing bindings is pointless here but the caching of the compiled algorithm already produces a noticeable speedup.

### running a mapping project
The mapping service then runs a triple loop.
* For each source repository
  * For each source repository entity
    * For each target attribute
      * it makes a call to the algorithm service

This also already benefits from the caching of the compiled algorithm but we can still significantly speed up the performance of running a mapping project if we delegate the inner loops to the algorithm service. That'll allow the algorithm service to cache the bindings between computations of the target attribute values for a single source row. 

I already took a brief look but the changes were nontrivial and don't benefit the Questionnaire use case. So I'd rather make that a separate issue and schedule it as a (probably high priority) bugfix.

Reply:
Thanks for the explanation. Could you create a story in the backlog for this enhancement?

Reply:
I've noticed that we're missing unit tests for script engine value conversion for attribute types BOOL, DATE_TIME, LONG. Lets create them and add in a separate PR.

Reply:
Created http://wiki.gcc.rug.nl/ticket/5830#ticket